### ISSUE TITLE
WebStyledLabel alignment issue

### ISSUE THREAD

**Issue by mgarin** [mgarin/weblaf#393]
Quoting @kovadam69

I have problems with WebStyledLabel. I use them in WebTable header to display multi-line column headers. I set in the cell renderer that the label should be centered, however it does not center the labels, all labels are left-aligned. If I add an icon to the left of the label, the ICON gets CENTERED, but the label remains on the left side:

![Bug preview](http://i.imgur.com/iORthoU.png)

Also if I move the cursor over the column, the text gets and endellipsys even if there are plenty of places. I tried to add border around the label, but the border is drawn in the entire column width and height.

![Bug preview](http://i.imgur.com/b7VrE8j.png)

I tried to debug, but it seems that the Painter calculates a proper X/Y coordinate for the label, but starts to paint from the X=0 instead of the calculated X=121 or so. I did not have time to dig deeper in the painting process, but something is wrong there (it was working with the WebMultilineLabel flawlessly, and I just changed the Object creation, and left all other code untouched)

**Comment by mgarin** [comment]
Comment from @kovadam69

I might found the problem for the WebStyledLabel issue. In StyledLabelPainter @line 719-735 given following code:
```
        else if ( horizontalAlignment == CENTER )
        {
            int leftMostX = 0;
            if ( horizontalTextPosition == SwingConstants.RIGHT && label.getIcon () != null )
            {
                leftMostX += label.getIcon ().getIconWidth () + label.getIconTextGap ();
            }
            if ( insets != null )
            {
                labelWidth -= insets.right + insets.left;
                leftMostX += insets.left;
            }
            if ( label.getIcon () != null && horizontalTextPosition != SwingConstants.CENTER )
            {
                labelWidth -= label.getIcon ().getIconWidth () + label.getIconTextGap ();
            }
            textX = leftMostX + ( labelWidth - paintWidth ) / 2;
        }
```
This will almost ensure, that textX is always 0, because:
leftMostX is initialized as 0.
there is no icon, so there is no addition.
the insets are 0, so do not affect leftMostX and labelWidth
At the end labelWidth equals paintWidth, so leftMostX is left 0, which is set to textX, making the label starts painting from 0.
Funny thing is, that variable leftAlignmentX contains the proper X position for the text to be printed (in this case 121, text width is 26 (3 letters))
May be in line 721 the leftMostX should be initialized with leftAlignmentX instead of 0?

**Comment by mgarin** [comment]
Comment from @kovadam69

I did some trials with fixing the StyledLabelPainter. The suggested change helped, but still there are huge issues with the alignment calculations. Not only for CENTER, but also for RIGHT alignment the coordinates calculated are wrong. Fixing the textX moved the text to the CENTER, but if I added an Icon, it looked strange after that, the gap was too big. Same for the RIGHT aligned text, with icon the text was painted over the icon. Probably the icon position is determined also wrong.
Also the text is not painted somehow to the entire place available, but only a portion of the place, and sometimes I get the ... at the end of the text even if there is plenty of room there...

This is what I fixed, which helped a bit, but the output is still wrong:
```
if ( horizontalAlignment == RIGHT )
        {
            paintWidth = thisLineEndX - textX;
            textX = thisLineEndX - labelWidth;
            if ( insets != null )
            {
                textX -= insets.right;
            }
            if ( label.getIcon () != null && horizontalTextPosition == SwingConstants.LEFT )
            {
                textX -= label.getIcon ().getIconWidth () + label.getIconTextGap ();
            }
        }
        else if ( horizontalAlignment == CENTER )
        {
            int leftMostX = leftAlignmentX;
            if ( horizontalTextPosition == SwingConstants.RIGHT && label.getIcon () != null )
            {
                leftMostX += label.getIcon ().getIconWidth () + label.getIconTextGap ();
            }
            if ( insets != null )
            {
                labelWidth -= insets.right + insets.left;
                leftMostX += insets.left;
            }
            if ( label.getIcon () != null && horizontalTextPosition != SwingConstants.CENTER )
            {
                labelWidth -= label.getIcon ().getIconWidth () + label.getIconTextGap ();
            }
            textX = leftMostX + ( labelWidth - paintWidth ) / 2;
        }
```

**Comment by mgarin** [comment]
Trick of reproducing this issue was found:
```java
label.setLineWrap ( true );
```
With this flag label (but not the icon) starts to ignore horizontal alignment and also has issues with preferred size. This should now be sufficient to provide a fix.

**Comment by mgarin** [comment]
This is now fixed ( commits 81ae192 and 698fb58 ) along with some major internal improvements for all text-containing components in general.
